File: C++/Boundary_traversal_of_tree.cpp
/*
Idea:
finding left boundary nodes, then all bottom boundary leaf nodes, then reverse right boundary nodes.
*/

//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;
#define MAX_HEIGHT 100000

// Tree Node
struct Node
{
    int data;
    Node* left;
    Node* right;
};

// Utility function to create a new Tree Node
Node* newNode(int val)
{
    Node* temp = new Node;
    temp->data = val;
    temp->left = NULL;
    temp->right = NULL;

    return temp;
}


// Function to Build Tree
Node* buildTree(string str)
{
    // Corner Case
    if (str.length() == 0 || str[0] == 'N')
        return NULL;

    // Creating vector of strings from input
    // string after spliting by space
    vector<string> ip;

    istringstream iss(str);
    for (string str; iss >> str; )
        ip.push_back(str);

    // Create the root of the tree
    Node* root = newNode(stoi(ip[0]));

    // Push the root to the queue
    queue<Node*> queue;
    queue.push(root);

    // Starting from the second element
    int i = 1;
    while (!queue.empty() && i < ip.size()) {

        // Get and remove the front of the queue
        Node* currNode = queue.front();
        queue.pop();

        // Get the current node's value from the string
        string currVal = ip[i];

        // If the left child is not null
        if (currVal != "N") {

            // Create the left child for the current node
            currNode->left = newNode(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->left);
        }

        // For the right child
        i++;
        if (i >= ip.size())
            break;
        currVal = ip[i];

        // If the right child is not null
        if (currVal != "N") {

            // Create the right child for the current node
            currNode->right = newNode(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->right);
        }
        i++;
    }

    return root;
}









// } Driver Code Ends


class Solution {
public:

    // function to check if a node is a leaf node or not
    bool isLeaf(Node* node)
    {
        if (node->left == NULL && node->right == NULL)
            return true;
        return false;
    }

    // function to find all leaf nodes at the bottom of the tree
    void leafNode(Node* root, vector<int>& ans)
    {
        if (!root)
            return;

        if (root->left)
            leafNode(root->left, ans);

        if (isLeaf(root))
            ans.push_back(root->data);

        if (root->right)
            leafNode(root->right, ans);
    }

    // function to find all left nodes
    void leftSubTree(Node* root, vector<int>& ans)
    {
        if (!root)
            return;

        // Check if the current node is not a leaf and add it to the answer
        if (!isLeaf(root))
            ans.push_back(root->data);

        // Traverse the left subtree first
        if (root->left)
            leftSubTree(root->left, ans);

        // If there is no left child, go to the right child
        else if (root->right)
            leftSubTree(root->right, ans);
    }

    // function to find all right nodes
    void rightSubTree(Node* root, vector<int>& ans)
    {
        if (!root)
            return;

        // Traverse the right subtree first
        if (root->right)
            rightSubTree(root->right, ans);

        // If there is no right child, go to the left child
        else if (root->left)
            rightSubTree(root->left, ans);

        // Check if the current node is not a leaf and add it to the answer
        if (!isLeaf(root))
            ans.push_back(root->data);
    }

    // Function to get the boundary traversal of the binary tree
    vector <int> boundary(Node* root)
    {
        vector<int> ans;
        ans.push_back(root->data); // Add the root data to the answer

        // Traverse the left subtree
        leftSubTree(root->left, ans);

        // Traverse the leaf nodes on the left and right subtrees
        leafNode(root->left, ans);
        leafNode(root->right, ans);

        // Traverse the right subtree
        rightSubTree(root->right, ans);

        return ans; // Return the final boundary traversal
    }
};


//{ Driver Code Starts.

/* Driver program to test size function*/

int main() {
    int t;
    string tc;
    getline(cin, tc);
    t = stoi(tc);
    while (t--)
    {
        string s, ch;
        getline(cin, s);
        Node* root = buildTree(s);
        Solution ob;
        vector <int> res = ob.boundary(root);
        for (int i : res) cout << i << " ";
        cout << endl;
    }
    return 0;
}
// } Driver Code Ends

File: Python/Boundary_traversal_of_tree.py
"""
Idea:
finding left boundary nodes, then all bottom boundary leaf nodes, then reverse right boundary nodes.
"""


# User function Template for python3


class Solution:
    # function to check if a node is a leaf node or not
    def isLeaf(self, node):
        if not node.left and not node.right:
            return True
        return False

    # function to find all leaf nodes at the bottom of the tree
    def leafNode(self, root, ans):
        if not root:
            return
        if root.left:
            self.leafNode(root.left, ans)
        if self.isLeaf(root):
            ans.append(root.val)
        if root.right:
            self.leafNode(root.right, ans)

    # function to find all left nodes
    def leftSubTree(self, root, ans):
        if not root:
            return
        if not self.isLeaf(root):
            ans.append(root.val)
        if root.left:
            self.leftSubTree(root.left, ans)
        elif root.right:
            self.leftSubTree(root.right, ans)

    # function to find all right nodes
    def rightSubTree(self, root, ans):
        if not root:
            return
        if root.right:
            self.rightSubTree(root.right, ans)
        elif root.left:
            self.rightSubTree(root.left, ans)
        if not self.isLeaf(root):
            ans.append(root.val)

    # Function to get the boundary traversal of the binary tree
    def boundaryOfBinaryTree(self, root):
        if not root:
            return []
        ans = [root.val]  # Add the root data to the answer
        # Traverse the left subtree
        self.leftSubTree(root.left, ans)
        # Traverse the leaf nodes on the left and right subtrees
        self.leafNode(root.left, ans)
        self.leafNode(root.right, ans)
        # Traverse the right subtree
        self.rightSubTree(root.right, ans)
        return ans  # Return the final boundary traversal


# Code here


# {
# Driver Code Starts
# Initial Template for Python 3

# function should return a list containing the boundary view of the binary tree
# {
#  Driver Code Starts
import sys

import sys

sys.setrecursionlimit(100000)
from collections import deque


# Tree Node
class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None


# Function to Build Tree
def buildTree(s):
    # Corner Case
    if (len(s) == 0 or s[0] == "N"):
        return None

    # Creating list of strings from input
    # string after spliting by space
    ip = list(map(str, s.split()))

    # Create the root of the tree
    root = Node(int(ip[0]))
    size = 0
    q = deque()

    # Push the root to the queue
    q.append(root)
    size = size + 1

    # Starting from the second element
    i = 1
    while (size > 0 and i < len(ip)):
        # Get and remove the front of the queue
        currNode = q[0]
        q.popleft()
        size = size - 1

        # Get the current node's value from the string
        currVal = ip[i]

        # If the left child is not null
        if (currVal != "N"):
            # Create the left child for the current node
            currNode.left = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.left)
            size = size + 1
        # For the right child
        i = i + 1
        if (i >= len(ip)):
            break
        currVal = ip[i]

        # If the right child is not null
        if (currVal != "N"):
            # Create the right child for the current node
            currNode.right = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.right)
            size = size + 1
        i = i + 1
    return root


if __name__ == "__main__":
    t = int(input())
    for _ in range(0, t):
        s = input()
        root = buildTree(s)
        obj = Solution()
        res = obj.printBoundaryView(root)
        for i in res:
            print(i, end=" ")
        print('')
# } Driver Code Ends
